### PR TITLE
[SPARK-10737] [SQL] When using UnsafeRows, SortMergeJoin may return wrong results

### PR DESCRIPTION
**Pull request by yhuai**
https://issues.apache.org/jira/browse/SPARK-10737

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -1781,4 +1781,32 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
         Seq(Row(1), Row(1)))
     }
   }
+
+  test("SortMergeJoin returns wrong results when using UnsafeRows") {
+    // This test is for the fix of https://issues.apache.org/jira/browse/SPARK-10737.
+    // This bug will be triggered when Tungsten is enabled and there are multiple
+    // SortMergeJoin operators executed in the same task.
+    val confs =
+      "spark.sql.planner.sortMergeJoin" -> "true" ::
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Could these use the objects so there will be a compile error if we ever remove these options? (and then we can delete this test)

Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala [hunk 2]
```diff
@@ -82,6 +79,28 @@ case class SortMergeJoin(
...
+            // Otherwise, GenerateProjection will cause wrong results.
+            UnsafeProjection.create(leftKeys, left.output)
+          } else {
+            newProjection(leftKeys, left.output)
+          }
+        }
+
+        // The projection used to extract keys from input rows of the right child.
+        private[this] val rightKeyGenerator = {
+          if (isUnsafeMode) {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala]: As discussed offline, can we move this check into `newProjection`?

**Comment by yhuai** [comment]
jenkins test this please

**Comment by yhuai** [comment]
test this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala (hunk 2 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala]: Had another discussion. We will do the cleanup work in a follow up pr.

**Comment by marmbrus** [comment]
LGTM

**Comment by yhuai** [comment]
@JoshRosen Can you also take a look?

**Comment by JoshRosen** [comment]
LGTM.

**Comment by yhuai** [comment]
Thank you for reviewing this PR! I am merging it to master and branch 1.5.

**Pull request closed without merging**